vtStorAtaProtocol/Platform/Linux/ProtocolAtaPassThrough.cpp

```diff
@@ -27,164 +28,142 @@ namespace vtStor
+        cEssenseAta1 essense = cEssenseAta1::Reader(Essense);
+
+        //! TODO: magic code 60
+        InitializePassThroughDirect(essense.GetCommandCharacteristics(), essense.GetTaskFileExt(), essense.GetTaskFile(), DataBuffer, 60);
+
+        errorCode = IssuePassThroughDirectCommand(Handle.Handle);
+    } break;
+
+    default:
+    {
+        errorCode = eErrorCode::FormatNotSupported;
+    } break;
     }
-*/
-   /* void cAtaPassThrough::InitializePassThroughDirect( const StorageUtility::Ata::sCommandCharacteristic& CommandCharacteristics, const StorageUtility::Ata::uTaskFileRegister& PreviousTaskFile, const StorageUtility::Ata::uTaskFileRegister& CurrentTaskFile, std::shared_ptr<cBufferInterface> DataBuffer, U32 TimeoutValueInSeconds )
+
+    return(errorCode);
+}
+
+eErrorCode cAtaPassThrough::IssuePassThroughDirectCommand(const int& FileDescriptor)
+{
+    assert(INVALID_HANDLE_VALUE != FileDescriptor);
+
+    U32 commandSuccessfulFlag = ioctl(FileDescriptor, SG_IO, &m_ATAPassThrough);
+    if (IOCTL_SG_IO_ERROR == commandSuccessfulFlag)
     {
-        m_AtaPassThrough.Length = sizeof( ATA_PASS_THROUGH_DIRECT );
-        m_AtaPassThrough.DataTransferLength = CommandCharacteristics.DataTransferLengthInBytes;
-        m_AtaPassThrough.TimeOutValue = TimeoutValueInSeconds;
-        if (nullptr != DataBuffer)
-        {
-            m_AtaPassThrough.DataBuffer = DataBuffer->ToDataBuffer();
-        }
-        m_AtaPassThrough.ReservedAsUchar = 0;
-        m_AtaPassThrough.ReservedAsUlong = 0;
-
-        InitializeFlags( CommandCharacteristics );
-        InitializeTaskFileInputRegisters( PreviousTaskFile, CurrentTaskFile );
+        //! ERROR: IOCTL SG_IO was not successful
+        return(eErrorCode::Io);
     }
 
-    void cAtaPassThrough::InitializeFlags( const StorageUtility::Ata::sCommandCharacteristic& AtaCommandCharacteristic )
+    return(eErrorCode::None);
+}
+
+void cAtaPassThrough::InitializePassThroughDirect(
+        const StorageUtility::Ata::sCommandCharacteristic& CommandCharacteristics,
+        const StorageUtility::Ata::uTaskFileRegister& PreviousTaskFile,
+        const StorageUtility::Ata::uTaskFileRegister& CurrentTaskFile,
+        std::shared_ptr<cBufferInterface> DataBuffer,
+        U32 TimeoutValueInSeconds
+        )
+{
+
+    memset(&m_ATAPassThrough, 0, sizeof(m_ATAPassThrough));
+
+    //! TODO: Set up for Sense Data
+
+    m_ATAPassThrough.interface_id     = SG_IO_INTERFACE_ID;
+    m_ATAPassThrough.mx_sb_len        = MAX_SB_LEN;
+    m_ATAPassThrough.dxfer_len        = CommandCharacteristics.DataTransferLengthInBytes;
+    m_ATAPassThrough.timeout          = TimeoutValueInSeconds;
+    m_ATAPassThrough.cmd_len          = SG_CDB_16_LEN;
+
+    if (nullptr != DataBuffer)
+    {
+        m_ATAPassThrough.dxferp = DataBuffer->ToDataBuffer();
+    }
+
+    InitializeFlags(CommandCharacteristics);
+    InitializeCdbRegisters(PreviousTaskFile, CurrentTaskFile);
+
+    m_ATAPassThrough.cmdp = (U8*)&m_Cdb16;

```

Comment:
"The purpose of union is to save memory by using the same memory region for storing different objects at different times."
http://stackoverflow.com/questions/2310483/purpose-of-unions-in-c-and-c
We should not use for cast type. it can change structure for process.